admiral/pkg/clusters/routingpolicy_handler.go

```diff
@@ -15,51 +16,201 @@ import (
 )
 
 type RoutingPolicyHandler struct {
+	RemoteRegistry       *RemoteRegistry
+	ClusterID            string
+	RoutingPolicyService RoutingPolicyProcessor
+}
+
+func NewRoutingPolicyHandler(rr *RemoteRegistry, cId string, rpProcessor RoutingPolicyProcessor) *RoutingPolicyHandler {
+	return &RoutingPolicyHandler{RemoteRegistry: rr, ClusterID: cId, RoutingPolicyService: rpProcessor}
+}
+
+type RoutingPolicyProcessor interface {
+	ProcessAddOrUpdate(ctx context.Context, eventType admiral.EventType, newRP *v1.RoutingPolicy, oldRP *v1.RoutingPolicy, dependents map[string]string) error
+	ProcessDependency(ctx context.Context, eventType admiral.EventType, dependency *v1.Dependency) error
+	Delete(ctx context.Context, eventType admiral.EventType, routingPolicy *v1.RoutingPolicy) error
+}
+
+type RoutingPolicyService struct {
 	RemoteRegistry *RemoteRegistry
-	ClusterID      string
 }
 
-type routingPolicyCache struct {
-	// map of routing policies key=environment.identity, value: RoutingPolicy object
-	// only one routing policy per identity + env is allowed
-	identityCache map[string]*v1.RoutingPolicy
-	mutex         *sync.Mutex
+func (r *RoutingPolicyService) ProcessAddOrUpdate(ctx context.Context, eventType admiral.EventType, newRP *v1.RoutingPolicy, oldRP *v1.RoutingPolicy, dependents map[string]string) error {
+	var err error
+	for _, remoteController := range r.RemoteRegistry.remoteControllers {
+		if !common.DoRoutingPolicyForCluster(remoteController.ClusterID) {
+			log.Warnf(LogFormat, eventType, "routingpolicy", newRP.Name, remoteController.ClusterID, "RoutingPolicy disabled for cluster")
+			continue
+		}
+		for _, dependent := range dependents {
+			// Check if the dependent exists in this remoteCluster. If so, we create an envoyFilter with dependent identity as workload selector
+			if _, ok := r.RemoteRegistry.AdmiralCache.IdentityClusterCache.Get(dependent).Copy()[remoteController.ClusterID]; ok {
+				_, err1 := createOrUpdateEnvoyFilter(ctx, remoteController, newRP, eventType, dependent, r.RemoteRegistry.AdmiralCache)
+				if err1 != nil {
+					log.Errorf(LogErrFormat, eventType, "routingpolicy", newRP.Name, remoteController.ClusterID, err)
+					err = common.AppendError(err, err1)
+				} else {
+					log.Infof(LogFormat, eventType, "routingpolicy	", newRP.Name, remoteController.ClusterID, "created envoyfilters")
+				}
+			}
+			id := common.GetRoutingPolicyIdentity(newRP)
+			if oldRP != nil {
+				oldEnv := common.GetRoutingPolicyEnv(oldRP)
+				r.RemoteRegistry.AdmiralCache.RoutingPolicyCache.Delete(id, oldEnv, oldRP.Name)
+			}
```

Comment:
Why are we deleting from the cache?

Reply:
We are deleting the old resource as part of the update event. Should we be retaining older resources?

Reply:
Since we dont have routing policies in ppd or production we may not need this. Plus cache ( in memory) gets populated every time admiral restarts.

Reply:
the reason i am deleting is if the `env` label is changed the cache will be corrupted which could cause problems for subsequent updates. ( ex: change is admiral.io/env label value )

Reply:
@adilfulara is it ok to refresh the map if there are error in the above loop? For example, what if all the calls to `createOrUpdateEnvoyFilter` return failure. Is it ok to mutate the map?

Reply:
done. good catch. fixed it.